### ISSUE TITLE
Cache is invalid when using containers environment variables

### ISSUE THREAD

**Issue by wnkz** [aws/aws-sam-cli#2885]
### Description:

Cache is always invalid when building with containers environment variables (`-e`, `--container-env-var`)

### Steps to reproduce:

- Create a any sample app with `sam init`
- Build multiple times with `sam build -cu -e FOO=BAR`

### Observed result:

- "Cache is invalid" message
- Function's hash is changing

```
❯ sam build -cu -e FOO=BAR
Starting Build use cache
Starting Build inside a container
Cache is invalid, running build and copying resources to function build definition of b89956ad-7a89-401a-98b5-68dad76c9c2d
Building codeuri: /private/var/folders/bf/f_z6y6n959193x3csh_pcb0m0000gn/T/tmp.Zjp9bgzU/sam-app/hello_world runtime: python3.8 metadata: {} functions: ['HelloWorldFunction']

Fetching public.ecr.aws/sam/build-python3.8:latest Docker container image......
Mounting /private/var/folders/bf/f_z6y6n959193x3csh_pcb0m0000gn/T/tmp.Zjp9bgzU/sam-app/hello_world as /tmp/samcli/source:ro,delegated inside runtime container

Build Succeeded

Built Artifacts  : .aws-sam/build
Built Template   : .aws-sam/build/template.yaml

Commands you can use next
=========================
[*] Invoke Function: sam local invoke
[*] Deploy: sam deploy --guided

Running PythonPipBuilder:ResolveDependencies
Running PythonPipBuilder:CopySource
❯ sam build -cu -e FOO=BAR
Starting Build use cache
Starting Build inside a container
Cache is invalid, running build and copying resources to function build definition of fb423804-b45f-431a-9eca-e5d58909cbe6
Building codeuri: /private/var/folders/bf/f_z6y6n959193x3csh_pcb0m0000gn/T/tmp.Zjp9bgzU/sam-app/hello_world runtime: python3.8 metadata: {} functions: ['HelloWorldFunction']

Fetching public.ecr.aws/sam/build-python3.8:latest Docker container image......
Mounting /private/var/folders/bf/f_z6y6n959193x3csh_pcb0m0000gn/T/tmp.Zjp9bgzU/sam-app/hello_world as /tmp/samcli/source:ro,delegated inside runtime container

Build Succeeded

Built Artifacts  : .aws-sam/build
Built Template   : .aws-sam/build/template.yaml

Commands you can use next
=========================
[*] Invoke Function: sam local invoke
[*] Deploy: sam deploy --guided

Running PythonPipBuilder:ResolveDependencies
Running PythonPipBuilder:CopySource
```

### Expected result:

"Valid cache found" message on second run

```
❯ sam build -cu
Starting Build use cache
Starting Build inside a container
Cache is invalid, running build and copying resources to function build definition of 72cbac78-847c-4b54-9953-5fe0d300b926
Building codeuri: /private/var/folders/bf/f_z6y6n959193x3csh_pcb0m0000gn/T/tmp.Zjp9bgzU/sam-app/hello_world runtime: python3.8 metadata: {} functions: ['HelloWorldFunction']

Fetching public.ecr.aws/sam/build-python3.8:latest Docker container image......
Mounting /private/var/folders/bf/f_z6y6n959193x3csh_pcb0m0000gn/T/tmp.Zjp9bgzU/sam-app/hello_world as /tmp/samcli/source:ro,delegated inside runtime container

Build Succeeded

Built Artifacts  : .aws-sam/build
Built Template   : .aws-sam/build/template.yaml

Commands you can use next
=========================
[*] Invoke Function: sam local invoke
[*] Deploy: sam deploy --guided

Running PythonPipBuilder:ResolveDependencies
Running PythonPipBuilder:CopySource
❯ sam build -cu
Starting Build use cache
Starting Build inside a container
Valid cache found, copying previously built resources from function build definition of 72cbac78-847c-4b54-9953-5fe0d300b926

Build Succeeded

Built Artifacts  : .aws-sam/build
Built Template   : .aws-sam/build/template.yaml

Commands you can use next
=========================
[*] Invoke Function: sam local invoke
[*] Deploy: sam deploy --guided
```

### Additional environment details (Ex: Windows, Mac, Amazon Linux etc)

1. OS: macOS 11.3.1
2. `sam --version`: 1.23.0
3. AWS region: eu-west-1

[debug.log](https://github.com/aws/aws-sam-cli/files/6493890/debug.log)

**Comment by wnkz** [comment]
This is very painful for "large" projects ; any plans / timeline on a fix?

**Comment by wnkz** [comment]
Tracked it down a bit more, it appears that when building with `-e`, those environment variables are stored in the build.toml file. On later builds, they are read again as if they where part of the `FunctionBuildDefinition ` so the function is treated as unique because the real definition (template.yml) does not have those.

https://github.com/aws/aws-sam-cli/blob/199ded1901964ed17023e9d5ed5682ff77ae3dfd/samcli/lib/build/build_graph.py#L189-L207

https://github.com/aws/aws-sam-cli/blob/199ded1901964ed17023e9d5ed5682ff77ae3dfd/samcli/lib/build/build_graph.py#L442-L469

**Comment by qingchm** [comment]
This fix is now released with our newest version 1.25.0! Please update the CLI version to use cached build with container env vars! Closing this ticket, let me know if you have any questions ;)
